gateway/dataimpl/server_v1/xdcrserver.go

```diff
@@ -0,0 +1,849 @@
+					zap.Uint16("vbucket_id", vbId),
+					zap.Error(err),
+				)
+
+				waitCh <- err
+				return
+			}
+
+			// we ignore send errors here as there isn't much point in trying to cancel
+			// the running operation anyways, so we just let them all fail.
+			_ = out.Send(&internal_xdcr_v1.GetVbucketInfoResponse{
+				Vbuckets: []*internal_xdcr_v1.GetVbucketInfoResponse_VbucketState{vbState},
+			})
+			waitCh <- nil
+		}(vbId)
+	}
+
+	for range fetchVbs {
+		err := <-waitCh
+		if err != nil {
+			return s.errorHandler.NewGenericStatus(err).Err()
+		}
+	}
+
+	return nil
+}
+
+func (s *XdcrServer) WatchCollections(in *internal_xdcr_v1.WatchCollectionsRequest, out internal_xdcr_v1.XdcrService_WatchCollectionsServer) error {
+	ctx := out.Context()
+	bucketAgent, oboUser, errSt := s.authHandler.GetHttpOboAgent(ctx, &in.BucketName)
+	if errSt != nil {
+		return errSt.Err()
+	}
+
+	var latestManifestUid uint64 = 0
+
+	for {
+		manifest, err := bucketAgent.GetCollectionManifest(out.Context(), &cbmgmtx.GetCollectionManifestOptions{
+			BucketName: in.BucketName,
+			OnBehalfOf: oboUser,
+		})
+		if err != nil {
+			return s.errorHandler.NewGenericStatus(err).Err()
+		}
+
+		manifestUid, _ := strconv.ParseUint(manifest.UID, 16, 32)
+		if manifestUid <= latestManifestUid {
+			select {
+			case <-time.After(5 * time.Second):
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+			continue
+		}
+
+		latestManifestUid = manifestUid
+
+		resp := &internal_xdcr_v1.WatchCollectionsResponse{
+			ManifestUid: uint32(manifestUid),
+		}
+		for _, scope := range manifest.Scopes {
+			scopeId, _ := strconv.ParseUint(scope.UID, 16, 32)
+
+			scopeOut := &internal_xdcr_v1.WatchCollectionsResponse_Scope{
+				ScopeId:   uint32(scopeId),
+				ScopeName: scope.Name,
+			}
+
+			for _, collection := range scope.Collections {
+				collectionId, _ := strconv.ParseUint(collection.UID, 16, 32)
+
+				collectionOut := &internal_xdcr_v1.WatchCollectionsResponse_Collection{
+					CollectionId:   uint32(collectionId),
+					CollectionName: collection.Name,
+				}
+
+				scopeOut.Collections = append(scopeOut.Collections, collectionOut)
+			}
+
+			resp.Scopes = append(resp.Scopes, scopeOut)
+		}
+
+		err = out.Send(resp)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (s *XdcrServer) GetDocument(
+	ctx context.Context,
+	in *internal_xdcr_v1.GetDocumentRequest,
+) (*internal_xdcr_v1.GetDocumentResponse, error) {
+	bucketAgent, oboUser, errSt := s.authHandler.GetMemdOboAgent(ctx, in.BucketName)
+	if errSt != nil {
+		return nil, errSt.Err()
+	}
+
+	errSt = s.checkKey(in.Key)
+	if errSt != nil {
+		return nil, errSt.Err()
+	}
+
+	if !in.IncludeContent && (in.IncludeXattrs != nil && *in.IncludeXattrs) {
+		return nil, status.New(codes.InvalidArgument, "IncludeXattrs cannot be true when IncludeContent is false").Err()
+	}
+
+	resp := &internal_xdcr_v1.GetDocumentResponse{}
+	var metaCas uint64
+	var dataCas uint64
+
+	for i := 0; i < 7; i++ {
```

Comment:
Why are we doing this for i<7? Might be worth having a named constant to explain.

Reply:
It's just a prime number amount of retries.  I could probably put a "MAX_RETRIES" constant there, but it seemed unneccessary.